File: js/godot/pckFile.js
import sortPaths from 'sort-paths'
import {BinaryReader} from "../binaryReader";
import aes from 'aes-js';

class PckEntry {
  constructor(reader, fmtVersion, fileOffset, origOffset) {
    this.path = reader.readString();
    this.offset = fileOffset + Number(reader.readUInt64()) - origOffset;
    this.size = Number(reader.readUInt64());
    this.hash = reader.readGUID();

    this.flags = 0;
    if (fmtVersion >= 2) {
      this.flags = reader.readUInt32();
    }
  }
}

export class PckFile {
  constructor(reader, origOffset = 0) {
    if (reader.readChars(4) !== 'GDPC') {
      throw new Error('Not a Godot package!');
    }
    this.formatVersion = reader.readUInt32();
    this.major = reader.readUInt32();
    this.minor = reader.readUInt32();
    this.patch = reader.readUInt32();

    this.fileFlags = 0;
    this.fileOffset = 0;
    if (this.formatVersion >= 2) {
      this.fileFlags = reader.readUInt32();
      this.fileOffset = Number(reader.readUInt64());
    }

    reader.read(64);  // reserved

    this.numFiles = reader.readUInt32();

    if (this.fileFlags & 1) {
      throw new Error('Cannot load PCK with encrypted dir');
      // const md5 = reader.readGUID();
      // const length = reader.readUInt64();
      // const iv = reader.read(16);
      // let encData = reader.read(Number(length));
      // const key = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
      // console.log('keylen:', key.length)
      // const aesCfb = new aes.ModeOfOperation.cfb(key, iv);
      // console.log(aesCfb.decrypt(encData));
    }

    this.files = [];
    for (let i = 0; i < this.numFiles; i++) {
      this.files.push(new PckEntry(reader, this.formatVersion, this.fileOffset, origOffset));
    }
    this.reader = reader;

    this.pathOrder = sortPaths(this.files.map(f => f.path), '/');
  }

  getFile(path) {
    if (!path.startsWith('res://')) {
      path = 'res://' + path;
    }
    return this.files.filter(f => f.path === path)[0];
  }

  getData(file) {
    this.reader.seek(file.offset);
    return this.reader.read(file.size);
  }
}

File: src/godot/pck_file.rs
use std::sync::{Arc, Mutex, MutexGuard};
use anyhow::anyhow;
use bytes::{Buf, Bytes};
use crate::base::asset::{Asset, AssetMetadata};
use crate::base::asset::bundle::BundleFile;
use crate::base::asset::provider::{AssetProvider, ProviderMetadata};
use crate::base::asset::types::AssetType;

#[derive(Debug, Clone)]
pub(crate) struct PckEntry {
    path: String,
    offset: usize,
    size: usize,
    hash: Bytes,
    flags: u32
}

impl PckEntry {
    pub(crate) fn from_bytes(data: &mut Bytes, format_version: i32, file_offset: usize, original_offset: usize) -> Self {
        let path = data.get_string();
        let offset = file_offset + data.get_u64_le() as usize - original_offset;
        let size = data.get_u64_le() as usize;
        let hash = data.slice(0..16);
        data.advance(16);
        let flags = if format_version >= 2 { data.get_u32_le() } else { 0 };

        PckEntry {
            path,
            offset,
            size,
            hash,
            flags,
        }
    }
}

#[derive(Debug, Clone)]
pub(crate) struct PckFile {
    format_version: i32,
    major: i32,
    minor: i32,
    patch: i32,
    file_flags: i32,
    file_offset: usize,
    files: Vec<PckEntry>
}

impl PckFile {
    pub(crate) fn new(data: &mut Bytes) -> anyhow::Result<Self> {
        if data.get_chars(4) != "GDPC" {
            return Err(anyhow!("Not a Godot PCK"));
        }
        let format_version = data.get_i32_le();
        let major = data.get_i32_le();
        let minor = data.get_i32_le();
        let patch = data.get_i32_le();

        let mut file_flags = 0;
        let mut file_offset = 0;
        if format_version >= 2 {
            file_flags = data.get_i32_le();
            file_offset = data.get_u64_le() as usize;
        }

        data.advance(64);  // reserved

        let num_files = data.get_i32_le();

        if file_flags & 1 != 0 {
            return Err(anyhow!("Cannot load encrypted PCK!"));
        }

        let mut files = Vec::new();
        for _ in 0..num_files {
            files.push(PckEntry::from_bytes(data, format_version, file_offset, 0));
        }

        Ok(PckFile {
            format_version,
            major,
            minor,
            patch,
            file_flags,
            file_offset,
            files,
        })
    }
}

impl BundleFile for PckFile {
    fn list_providers(&self) -> Vec<ProviderMetadata> {
        self.files.iter().map(|f| ProviderMetadata {
            name: f.path.clone(),
            id: f.path.clone(),
        }).collect()
    }

    fn list_blobs(&self) -> Vec<ProviderMetadata> {
        todo!()
    }

    fn get_provider(&mut self, id: String) -> Option<Arc<Box<dyn AssetProvider>>> {
        todo!()
    }

    fn get_blob(&mut self, id: String) -> Option<Bytes> {
        todo!()
    }
}
